### DISCUSSION TITLE
rotation and yaw

### DISCUSSION THREAD

**Discussion by fangjir** [bitcraze/discussions#1321] [Q&A]
i have two question about the crazyflie 2.1
1. what is the rotation order?
2. i consider the value of yaw via log stabilizer.yaw ,the value is often 0 when the crazyflie power on No matter how I place it
i want to know whether this is right, and when i use send_setpoint command,how to use the code about the yaw?

**Comment by ToveRumar** [comment]
Hi!
Thank you for your questions!  

>     1. what is the rotation order?
I would need some more information to answer this question. How are you using the crazyflie? 
How is it controlled, and are you using any positioning system?
> 
>     2. i consider the value of yaw via log stabilizer.yaw ,the value is often 0 when the crazyflie power on No matter how I place it
>        i want to know whether this is right, and when i use send_setpoint command,how to use the code about the yaw?

Unless you  are using a positioning system yaw will always be 0 on startup. 
The yaw relative to body frame will in fact always be relative to 0, which is the value on startup. If you are using a positioning system the yaw might be transformed to another frame (based on your setup).   

Im not exactly sure what you mean by 
>   and when i use send_setpoint command,how to use the code about the yaw?

Could you please specify your use case a bit more and what you would like to achieve?

**Comment by fangjir** [comment]
if yaw is always be 0 on startup ,the crazyflie how to get the rotation from body frame to world frame(ENU)？

**Reply by fangjir** [reply]
@ToveRumar

**Reply by ToveRumar** [reply]
This is possible using a positioning system, could you please supply some more information on your setup. 

If there is no positioning system the body frame == world frame.  
How this rotation is achieved depends on your setup.

**Reply by Residualstress** [reply]
Hi, Tove Rumar! Actually, I was confused with the yaw angle recently. I want to use the value of the rotation from body frame to world frame. So I set up the CF Bolt with Loco position system. But I found the  `stateEstimate.yaw` would change when I rotated the drone and decrease to 0 when it stay still. 

So how can I get the heading angle in Loco position system?

**Reply by ToveRumar** [reply]
Hi!

It is not possible to get the heading angle from the Loco positioning system. 
The heading will be relative to the heading when starting up. 

To get this correct in the system , you should start the Crazyflie facing forward along the x-axis (towards positive).  This will make the yaw angle align with the positioning system (see; **https://www.bitcraze.io/documentation/system/platform/cf2-coordinate-system/** for info). 

If you want the starting position to be aligned differently you need to configure the system with the new starting attitude using this script; https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/positioning/initial_position.py

**Reply by Residualstress** [reply]
Thanks for your reply. Even if I start the CF facing forward along the x-aixs to make the yaw angle equal to heading angle, but it would decrease to ZERO when the drone is just hover.

For example, if I rotate the drone in 30 degree, the yaw would be 30 degree. But this value will decrease to ZERO when the heading direction doesn't change. So I can't take the `stateEstimate.yaw` as heading angle when drone's heading direction keep same.

**Reply by knmcguire** [reply]
There is a difference between the Kalman filter if it is on the ground compared to when it is is flying, but if you say this even happens in hover that is a bit concerning. We will need to investigate this again.

There was an issue open on this, but we will look into it again it seems https://github.com/bitcraze/crazyflie-firmware/issues/870

**Comment by gemenerik** [comment]
Hi @Residualstress,

It's been a while since this discussion, but if you still remember, could you clarify the issue you experienced? Specifically, was the yaw estimate converging to zero during steady hover—in flight, using the Loco Positioning System?

For example, if you take off with zero yaw, then rotate to 30 degrees yaw, does the state estimate gradually shift back to 0 degrees while the drone physically remains at 30? Or does the drone itself physically rotate back to 0 degrees when you expect it to stay at 30?

If you still have any code where you observed this behavior, would you be able to share it? In hindsight, we should have asked these questions back then, but better late than never!

Thanks in advance for any insights you can share.

**Reply by Residualstress** [reply]
Hi, Rik. I don't remember the detail of the experiment since I couldn't make it. But I can redo two simple experiments to observe the results.
1. Control the Crazy Bolt using Joystick 
2. Control the Crazy Bolt using App Layer

But I'm not in the lab this week. I'll let you know the results next week when I'm done.

I also have a question is that can I set the setpoint with control the velocity and yaw at the same time?
```
static void setHoverSetpoint(setpoint_t *setpoint, float vx, float yaw)
{

  setpoint->mode.x = modeVelocity;
  setpoint->velocity.x = vx;
  setpoint->mode.pitch = modeAbs;

  
  setpoint->mode.yaw = modeAbs;
  setpoint->attitude.yaw= yaw;

  setpoint->velocity_body = true;
}
```
From experiment results, they couldn't be set at the same time.

**Reply by gemenerik** [reply]
In the forum thread https://forum.bitcraze.io/viewtopic.php?t=3843, a very similar approach was successfully implemented, as noted in the edit of the post. However, for simply testing a yaw setpoint, you don’t necessarily need to do it in this way. Using the commander interface is a simpler way.

If it doesn’t work as expected, let me know and I’ll look into it further.
